### PR TITLE
support flash attention with sparse mask

### PR DESCRIPTION
**Pull request by GuoxiaWang**
### PR types

New features
### PR changes

APIs
### Description

support flash attention with sparse mask
Pcard-73145

### CODE REVIEW

File: python/paddle/nn/functional/flash_attention.py [hunk 1]
```diff
@@ -559,3 +559,141 @@ def scaled_dot_product_attention(
...
+    return_softmax_lse=False,
+    return_seed_offset=False,
+    training=True,
+    name=None,
+):
+    r"""
+    The equation is:
+    .. math::
+        result=softmax(\frac{ Q * K^T }{\sqrt{d}}) * V
+    where : ``Q``, ``K``, and ``V`` represent the three input parameters of the attention module.
```
Comment (sunzhongkai588) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
    The equation is:
    
    .. math::
        result=softmax(\frac{ Q * K^T }{\sqrt{d}}) * V
        
    where : ``Q``, ``K``, and ``V`` represent the three input parameters of the attention module.
```

File: python/paddle/nn/functional/flash_attention.py [hunk 2]
```diff
@@ -559,3 +559,141 @@ def scaled_dot_product_attention(
...
+    The dimensions of the three parameters are the same.
+    ``d`` represents the size of the last dimension of the three parameters.
+    Warning:
+        This API only supports inputs with dtype float16 and bfloat16.
+    Args:
```
Comment (sunzhongkai588) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
    ``d`` represents the size of the last dimension of the three parameters.
    
    Warning:
        This API only supports inputs with dtype float16 and bfloat16.
        
    Args:
```

File: python/paddle/nn/functional/flash_attention.py [hunk 3]
```diff
@@ -559,3 +559,141 @@ def scaled_dot_product_attention(
...
+                        it can set `attn_mask_start_row` for performance improvement.
+                        The default value is 0.
+        dropout_p(float): The dropout ratio.
+        is_causal(bool): Whether enable causal mode.
+        training(bool): Whether it is in the training phase.
+        name(str, optional): The default value is None. Normally there is no need for user
+                        to set this property. For more information, please refer to
+                        :ref:`api_guide_Name`.
+    Returns:
+        out(Tensor): The attention tensor.
```
Comment (sunzhongkai588) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
        out(Tensor), The attention tensor.
```

File: python/paddle/nn/functional/flash_attention.py [hunk 4]
```diff
@@ -559,3 +559,141 @@ def scaled_dot_product_attention(
...
+                    4-D tensor with shape: [batch_size, seq_len, num_heads, head_dim].
+                    The dtype can be float16 or bfloat16.
+    Examples:
+        .. code-block:: python
+            >>> # doctest: +SKIP('bfloat need V100 compile')
```
Comment (sunzhongkai588) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
        .. code-block:: python
        
            >>> # doctest: +SKIP('bfloat need V100 compile')
```

**Review by sunzhongkai588 — COMMENTED** [review]
修改文档以符合规范

**Review by sunzhongkai588 — APPROVED** [review]
LGTM for docs

**Review by YuanRisheng — APPROVED** [review]

**Review by kolinwei — APPROVED** [review]

File: paddle/phi/api/yaml/backward.yaml [hunk 5]
```diff
@@ -859,6 +859,17 @@
     func : flash_attn_unpadded_grad
     data_type: q
 
+- backward_op : flash_attn_with_sparse_mask_grad
+  forward : flash_attn_with_sparse_mask (Tensor q, Tensor k, Tensor v, Tensor attn_mask_start_row_indices, Tensor fixed_seed_offset, float dropout = 0.0, bool causal = false, int attn_mask_start_row = 0, bool return_softmax = false, bool is_test = false, str rng_name = "") -> Tensor(out), Tensor(softmax), Tensor(softmax_lse), Tensor(seed_offset)
+  args : (Tensor q, Tensor k, Tensor v, Tensor attn_mask_start_row_indices, Tensor out, Tensor softmax_lse, Tensor seed_offset, Tensor out_grad, float dropout = 0.0, bool causal = false, int attn_mask_start_row = 0)
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/backward.yaml:864]: 后续该算子若要支持自动并行，请添加切分推导规则

**Review by heavyrain-lzy — APPROVED** [review]
LGTM for YAML

**Review by zhangbo9674 — APPROVED** [review]

File: paddle/phi/api/yaml/backward.yaml (hunk 5 above)
Comment (GuoxiaWang) [reply paddle/phi/api/yaml/backward.yaml:864]: 好的

**Comment by jeff41404** [comment]
API is fine, please attach the Chinese document PR link in description above.

**Comment by deepllz** [comment]
> API is fine, please attach the Chinese document PR link in description above.

添加相关中文文档：https://github.com/PaddlePaddle/docs/pull/6554

**Review by jeff41404 — APPROVED** [review]
LGTM for API

**Review by sneaxiy — APPROVED** [review]

**Pull request merged**